contracts/core/Registry.sol

```diff
@@ -1,47 +1,174 @@
         string name;
-        string metadata; // @todo update to MetaPtr
+        Metadata metadata;
         address attestationAddr;
     }
 
-    // getter for projects mapping
-    function identities(uint _identityId) external view returns (IdentityDetails memory) {
+    /// @notice identityId -> IdentityDetails
+    mapping(uint => IdentityDetails) public identities;
 
-    }
+    // Events
+    event IdentityCreated(uint indexed identityId, string name, Metadata metadata, address attestationAddr);
+    event IdentityNameUpdated(uint indexed identityId, string name);
+    event IdentityMetadataUpdated(uint indexed identityId, Metadata metadata);
+
+    /// @notice Creates a new identity
+    /// @dev This will also set the attestation address generated from msg.sender and name
+    function createIdentity(
+        string memory name,
+        Metadata memory metadata,
+        address[] memory _owners,
+        address[] memory _members
+    ) external returns (uint256) {
+
+        IdentityDetails memory identityDetails = IdentityDetails(
+            name,
+            metadata,
+            _generateAttestationAddr(identityId, name)
+        );
 
-    // create a new project with metadata
-    // hash some unique value to create the identityId - is there enough immutable to do this or just use incrementer?
-    // @todo if owners were immutable, we could make it a merkle proof of addresses and prove ownership that way? probably overkill
-    // sets roles so that all owners are owners of the project
-    // attestationAddr = addr(hash(id, name, owner))
-    // OR deploy right away with CREATE2 as proxy so they can do anything (or just forward ERC20s and ETH)
-    // id => IdentityDetails
-    function createIdentity(IdentityDetails memory _identityDetails, address[] memory _owners) external returns (uint256) {
+        identities[identityId] = identityDetails;
 
+        // generate roles
+        bytes32 ownerRole = _generateRole(identityId, RoleType.OWNER);
+        bytes32 memberRole = _generateRole(identityId, RoleType.MEMBER);
+
+        // assign roles
+        for(uint i = 0; i < _owners.length; i++) {
+            _grantRole(ownerRole, _owners[i]);
+        }
+        for(uint i = 0; i < _members.length; i++) {
+            _grantRole(memberRole, _members[i]);
+        }
+
+        emit IdentityCreated(identityId, name, metadata, identityDetails.attestationAddr);
+
+        identityId++;
+        
+        return identityId;
     }
 
-    // updates the name of the identity
-    // note that this will also change the attestation address, since it's a hash that includes the name
+    /// @notice Updates the name of the identity. 
+    /// @param _identityId The identityId of the identity
+    /// @param _name The new name of the identity
+    /// @dev Only owner can update the name. Also updated the attestation address
     function updateIdentityName(uint _identityId, string memory _name) external {
```

Comment:
Do we want to emit and/or return the new address?

Reply:
Good question, I don't see any harm in return data as well. We could add a return like so:

```javascript
function updateIdentityName(
    uint _identityId,
    string memory _name
) external returns (address) {
    bytes32 ownerRole = _generateRole(identityId, RoleType.OWNER);

    if (!hasRole(ownerRole, msg.sender)) {
        revert NO_ACCESS_TO_ROLE();
    }

    identities[_identityId].name = _name;
    identities[_identityId].ownerIdentifier = _generateownerIdentifier(
        _name,
        msg.sender
    );

    address ownerIdentifier = identities[_identityId].ownerIdentifier;

    emit IdentityNameUpdated(_identityId, _name, ownerIdentifier);

    return ownerIdentifier;
}
```

We could also update the emit:
```javascript
event IdentityNameUpdated(uint indexed identityId, string name, address newOwnerIdentifier);
```

Reply:
Can someone help me understand how role flexibility works here? Is it just these two roles (owner/member) or will identity owners have the flexibility to create additional roles?

Reply:
So far we have only defined these 2 roles, but we could look into a way for the `identityOwner` to create custom roles. I think that would be more complex and needs to be thought through a bit more.

We currently use an `enum` to pass when we create the role with the `msg.sender`. The function is also just used internally to this contract and not publicly callable.

```javascript
enum RoleType {
    OWNER,
    MEMBER
}

function _generateRole(
    uint _identityId,
    RoleType _roleType
) internal pure returns (bytes32 roleHash) {
    roleHash = keccak256(abi.encodePacked(_identityId, _roleType));
}
```

We can think through how we want to make this extendable.